File: kits/data-integration/providers/enricher_plugin/whisper_transcribe.ts
// COPF Data Integration Kit - Audio/video transcription enricher provider via Whisper
// Sends audio to Whisper API or runs locally, parses timestamped segments.

import * as https from 'https';
import { execFile } from 'child_process';
import { promisify } from 'util';
import { writeFileSync, readFileSync, unlinkSync } from 'fs';
import { tmpdir } from 'os';
import { join } from 'path';

const execFileAsync = promisify(execFile);

export const PROVIDER_ID = 'whisper_transcribe';
export const PLUGIN_TYPE = 'enricher_plugin';

export interface ContentItem {
  id: string;
  content: string;
  contentType: string;
  metadata?: Record<string, unknown>;
}

export interface EnricherConfig {
  model?: string;
  apiKey?: string;
  threshold?: number;
  options?: Record<string, unknown>;
}

export interface EnrichmentResult {
  fields: Record<string, unknown>;
  confidence: number;
  metadata?: Record<string, unknown>;
}

export interface SchemaRef {
  name: string;
  fields?: string[];
}

export interface CostEstimate {
  tokens?: number;
  apiCalls?: number;
  durationMs?: number;
}

interface TranscriptSegment {
  text: string;
  start: number;
  end: number;
  confidence: number;
}

type WhisperModel = 'tiny' | 'base' | 'small' | 'medium' | 'large';

function parseWhisperJsonOutput(jsonStr: string): TranscriptSegment[] {
  try {
    const data = JSON.parse(jsonStr);
    const segments: TranscriptSegment[] = (data.segments ?? []).map((seg: any) => ({
      text: (seg.text ?? '').trim(),
      start: seg.start ?? 0,
      end: seg.end ?? 0,
      confidence: 1.0 - (seg.no_speech_prob ?? 0),
    }));
    return segments;
  } catch {
    return [];
  }
}

function parseSrtOutput(srt: string): TranscriptSegment[] {
  const segments: TranscriptSegment[] = [];
  const blocks = srt.split(/\n\n+/).filter((b) => b.trim());

  for (const block of blocks) {
    const lines = block.split('\n');
    if (lines.length < 3) continue;

    // Parse timestamp line: "00:00:01,500 --> 00:00:04,200"
    const timeMatch = lines[1].match(
      /(\d{2}):(\d{2}):(\d{2})[,.](\d{3})\s*-->\s*(\d{2}):(\d{2}):(\d{2})[,.](\d{3})/,
    );
    if (!timeMatch) continue;

    const start =
      parseInt(timeMatch[1]) * 3600 +
      parseInt(timeMatch[2]) * 60 +
      parseInt(timeMatch[3]) +
      parseInt(timeMatch[4]) / 1000;
    const end =
      parseInt(timeMatch[5]) * 3600 +
      parseInt(timeMatch[6]) * 60 +
      parseInt(timeMatch[7]) +
      parseInt(timeMatch[8]) / 1000;

    const text = lines.slice(2).join(' ').trim();
    segments.push({ text, start, end, confidence: 0.85 });
  }

  return segments;
}

async function transcribeLocal(
  audioPath: string,
  modelSize: WhisperModel,
  language: string,
): Promise<TranscriptSegment[]> {
  const outputPath = join(tmpdir(), `copf_whisper_${Date.now()}`);

  try {
    await execFileAsync('whisper', [
      audioPath,
      '--model', modelSize,
      '--language', language,
      '--output_format', 'json',
      '--output_dir', tmpdir(),
      '--output_file', outputPath,
    ], { timeout: 300000, maxBuffer: 50 * 1024 * 1024 });

    const jsonContent = readFileSync(`${outputPath}.json`, 'utf-8');
    return parseWhisperJsonOutput(jsonContent);
  } finally {
    try { unlinkSync(`${outputPath}.json`); } catch { /* best-effort */ }
  }
}

async function transcribeApi(
  audioB64: string,
  apiKey: string,
  model: string,
  language: string,
): Promise<TranscriptSegment[]> {
  const audioBuffer = Buffer.from(audioB64, 'base64');
  const boundary = `----CopfBoundary${Date.now()}`;
  const parts: Buffer[] = [];

  // Build multipart/form-data for OpenAI Whisper API
  const addField = (name: string, value: string) => {
    parts.push(Buffer.from(
      `--${boundary}\r\nContent-Disposition: form-data; name="${name}"\r\n\r\n${value}\r\n`,
    ));
  };

  addField('model', model || 'whisper-1');
  addField('response_format', 'verbose_json');
  addField('timestamp_granularities[]', 'segment');
  if (language) addField('language', language);

  parts.push(Buffer.from(
    `--${boundary}\r\nContent-Disposition: form-data; name="file"; filename="audio.wav"\r\nContent-Type: audio/wav\r\n\r\n`,
  ));
  parts.push(audioBuffer);
  parts.push(Buffer.from(`\r\n--${boundary}--\r\n`));

  const body = Buffer.concat(parts);

  return new Promise((resolve, reject) => {
    const req = https.request({
      hostname: 'api.openai.com',
      path: '/v1/audio/transcriptions',
      method: 'POST',
      headers: {
        'Authorization': `Bearer ${apiKey}`,
        'Content-Type': `multipart/form-data; boundary=${boundary}`,
        'Content-Length': String(body.length),
      },
    }, (res) => {
      const chunks: Buffer[] = [];
      res.on('data', (chunk) => chunks.push(chunk));
      res.on('end', () => {
        try {
          const json = JSON.parse(Buffer.concat(chunks).toString());
          const segments = (json.segments ?? []).map((seg: any) => ({
            text: (seg.text ?? '').trim(),
            start: seg.start ?? 0,
            end: seg.end ?? 0,
            confidence: seg.avg_logprob != null ? Math.exp(seg.avg_logprob) : 0.85,
          }));
          resolve(segments);
        } catch (e) {
          reject(new Error(`Failed to parse Whisper API response: ${e}`));
        }
      });
    });
    req.on('error', reject);
    req.write(body);
    req.end();
  });
}

export class WhisperTranscribeEnricherProvider {
  async enrich(item: ContentItem, config: EnricherConfig): Promise<EnrichmentResult> {
    const modelSize = (config.options?.modelSize as WhisperModel) ?? 'base';
    const language = (config.options?.language as string) ?? 'en';
    const useApi = !!config.apiKey;

    let segments: TranscriptSegment[];

    if (useApi) {
      segments = await transcribeApi(item.content, config.apiKey!, config.model ?? 'whisper-1', language);
    } else {
      // Write audio to temp file for local processing
      const tmpPath = join(tmpdir(), `copf_audio_${item.id}_${Date.now()}.wav`);
      writeFileSync(tmpPath, Buffer.from(item.content, 'base64'));
      try {
        segments = await transcribeLocal(tmpPath, modelSize, language);
      } finally {
        try { unlinkSync(tmpPath); } catch { /* best-effort */ }
      }
    }

    const fullText = segments.map((s) => s.text).join(' ');
    const avgConfidence = segments.length > 0
      ? segments.reduce((sum, s) => sum + s.confidence, 0) / segments.length
      : 0;
    const totalDuration = segments.length > 0
      ? segments[segments.length - 1].end
      : 0;

    return {
      fields: {
        transcript: {
          text: fullText,
          segments,
        },
        word_count: fullText.split(/\s+/).filter(Boolean).length,
        duration_seconds: totalDuration,
      },
      confidence: avgConfidence,
      metadata: {
        provider: PROVIDER_ID,
        modelSize,
        language,
        segmentCount: segments.length,
        mode: useApi ? 'api' : 'local',
      },
    };
  }

  appliesTo(schema: SchemaRef): boolean {
    const audioSchemas = ['audio', 'video', 'podcast', 'recording', 'speech', 'media'];
    return audioSchemas.some((s) => schema.name.toLowerCase().includes(s));
  }

  costEstimate(item: ContentItem): CostEstimate {
    const sizeKb = Buffer.byteLength(item.content, 'base64') / 1024;
    // Approximate audio duration from file size: ~16KB/sec for WAV, ~2KB/sec for compressed
    const isCompressed = item.contentType.includes('mp3') || item.contentType.includes('ogg');
    const bytesPerSec = isCompressed ? 2048 : 16384;
    const estimatedDurationSec = (sizeKb * 1024) / bytesPerSec;
    // Whisper processes ~30 seconds per second of audio on medium hardware
    const processingMs = Math.max(1000, estimatedDurationSec * 33);

    return {
      durationMs: Math.round(processingMs),
      apiCalls: 1,
    };
  }
}

export default WhisperTranscribeEnricherProvider;

File: repertoire/data-integration/providers/enricher_plugin/WhisperTranscribe.swift
// Clef Data Integration Kit - Audio/video transcription enricher provider via Whisper
// Sends audio to Whisper API or runs locally, parses timestamped segments.

import Foundation

public let WhisperTranscribeProviderID = "whisper_transcribe"
public let WhisperTranscribePluginType = "enricher_plugin"

public struct TranscriptSegment {
    public let text: String
    public let start: Double
    public let end: Double
    public let confidence: Double
}

public final class WhisperTranscribeEnricherProvider {
    public init() {}

    public func enrich(item: ContentItem, config: EnricherConfig) async throws -> EnrichmentResult {
        let modelSize = (config.options?["modelSize"] as? String) ?? "base"
        let language = (config.options?["language"] as? String) ?? "en"
        let useApi = config.apiKey != nil && !config.apiKey!.isEmpty

        let segments: [TranscriptSegment]

        if useApi {
            segments = try await transcribeApi(
                audioB64: item.content, apiKey: config.apiKey!,
                model: config.model ?? "whisper-1", language: language
            )
        } else {
            segments = try await transcribeLocal(
                item: item, modelSize: modelSize, language: language
            )
        }

        let fullText = segments.map { $0.text }.joined(separator: " ")
        let avgConfidence: Double = segments.isEmpty ? 0.0 :
            segments.reduce(0.0) { $0 + $1.confidence } / Double(segments.count)
        let totalDuration = segments.last?.end ?? 0.0
        let wordCount = fullText.split(separator: " ").count

        let segmentDicts: [[String: Any]] = segments.map { seg in
            ["text": seg.text, "start": seg.start, "end": seg.end, "confidence": seg.confidence]
        }

        return EnrichmentResult(
            fields: [
                "transcript": ["text": fullText, "segments": segmentDicts] as [String: Any],
                "word_count": wordCount,
                "duration_seconds": totalDuration
            ],
            confidence: avgConfidence,
            metadata: [
                "provider": WhisperTranscribeProviderID,
                "modelSize": modelSize,
                "language": language,
                "segmentCount": segments.count,
                "mode": useApi ? "api" : "local"
            ]
        )
    }

    public func appliesTo(schema: SchemaRef) -> Bool {
        let audioSchemas = ["audio", "video", "podcast", "recording", "speech", "media"]
        let nameLower = schema.name.lowercased()
        return audioSchemas.contains { nameLower.contains($0) }
    }

    public func costEstimate(item: ContentItem) -> CostEstimate {
        let sizeKb = Double(item.content.count) * 3.0 / 4.0 / 1024.0
        let isCompressed = item.contentType.contains("mp3") || item.contentType.contains("ogg")
        let bytesPerSec: Double = isCompressed ? 2048.0 : 16384.0
        let estimatedDurationSec = (sizeKb * 1024.0) / bytesPerSec
        let processingMs = max(1000.0, estimatedDurationSec * 33.0)
        return CostEstimate(tokens: nil, apiCalls: 1, durationMs: Int(processingMs))
    }

    // MARK: - Local Transcription

    private func transcribeLocal(
        item: ContentItem, modelSize: String, language: String
    ) async throws -> [TranscriptSegment] {
        guard let audioData = Data(base64Encoded: item.content) else {
            throw EnricherError.parseError("Failed to decode base64 audio content")
        }

        let tmpPath = NSTemporaryDirectory() + "clef_audio_\(item.id)_\(ProcessInfo.processInfo.processIdentifier).wav"
        try audioData.write(to: URL(fileURLWithPath: tmpPath))
        defer { try? FileManager.default.removeItem(atPath: tmpPath) }

        let process = Process()
        process.executableURL = URL(fileURLWithPath: "/usr/bin/env")
        process.arguments = [
            "whisper", tmpPath,
            "--model", modelSize,
            "--language", language,
            "--output_format", "json",
            "--output_dir", NSTemporaryDirectory()
        ]

        let stdoutPipe = Pipe()
        let stderrPipe = Pipe()
        process.standardOutput = stdoutPipe
        process.standardError = stderrPipe

        try process.run()
        process.waitUntilExit()

        guard process.terminationStatus == 0 else {
            let errData = stderrPipe.fileHandleForReading.readDataToEndOfFile()
            let errMsg = String(data: errData, encoding: .utf8) ?? "Unknown error"
            throw EnricherError.processError("Whisper failed: \(errMsg)")
        }

        // Read the generated JSON output
        let baseName = (tmpPath as NSString).deletingPathExtension
        let jsonPath = NSTemporaryDirectory() + (baseName as NSString).lastPathComponent + ".json"
        let jsonData = try Data(contentsOf: URL(fileURLWithPath: jsonPath))
        defer { try? FileManager.default.removeItem(atPath: jsonPath) }

        return parseWhisperJson(data: jsonData)
    }

    // MARK: - API Transcription

    private func transcribeApi(
        audioB64: String, apiKey: String, model: String, language: String
    ) async throws -> [TranscriptSegment] {
        guard let audioData = Data(base64Encoded: audioB64) else {
            throw EnricherError.parseError("Failed to decode base64 audio")
        }

        let boundary = "----CopfBoundary\(Int(Date().timeIntervalSince1970 * 1000))"
        var body = Data()

        // Build multipart form data
        func appendField(name: String, value: String) {
            body.append("--\(boundary)\r\n".data(using: .utf8)!)
            body.append("Content-Disposition: form-data; name=\"\(name)\"\r\n\r\n".data(using: .utf8)!)
            body.append("\(value)\r\n".data(using: .utf8)!)
        }

        appendField(name: "model", value: model)
        appendField(name: "response_format", value: "verbose_json")
        appendField(name: "language", value: language)

        body.append("--\(boundary)\r\n".data(using: .utf8)!)
        body.append("Content-Disposition: form-data; name=\"file\"; filename=\"audio.wav\"\r\n".data(using: .utf8)!)
        body.append("Content-Type: audio/wav\r\n\r\n".data(using: .utf8)!)
        body.append(audioData)
        body.append("\r\n--\(boundary)--\r\n".data(using: .utf8)!)

        guard let url = URL(string: "https://api.openai.com/v1/audio/transcriptions") else {
            throw EnricherError.parseError("Invalid Whisper API URL")
        }
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("Bearer \(apiKey)", forHTTPHeaderField: "Authorization")
        request.setValue("multipart/form-data; boundary=\(boundary)", forHTTPHeaderField: "Content-Type")
        request.httpBody = body

        let (data, response) = try await URLSession.shared.data(for: request)
        guard let httpResponse = response as? HTTPURLResponse,
              (200...299).contains(httpResponse.statusCode) else {
            throw EnricherError.processError("Whisper API request failed")
        }

        return parseWhisperJson(data: data)
    }

    // MARK: - Parsing

    private func parseWhisperJson(data: Data) -> [TranscriptSegment] {
        guard let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let segments = json["segments"] as? [[String: Any]] else {
            return []
        }

        return segments.compactMap { seg in
            guard let text = seg["text"] as? String else { return nil }
            let start = (seg["start"] as? Double) ?? 0.0
            let end = (seg["end"] as? Double) ?? 0.0
            let confidence: Double
            if let avgLogprob = seg["avg_logprob"] as? Double {
                confidence = exp(avgLogprob)
            } else if let noSpeech = seg["no_speech_prob"] as? Double {
                confidence = 1.0 - noSpeech
            } else {
                confidence = 0.85
            }
            return TranscriptSegment(
                text: text.trimmingCharacters(in: .whitespaces),
                start: start, end: end, confidence: confidence
            )
        }
    }
}
